### PR TITLE
PSI integration

### LINKED ISSUE #13: Include PSI build in project
## Feature Description
Add relevant Python code from [PSI project](https://github.com/OpenMined/PSI) to the project.
Include relevant documentation for compiling/running in README

## Additional Context
Part of Epic #2

### LINKED ISSUE #22: Simple PSI demonstration
## Feature Description
Implement a function that takes two arrays of unique IDs and returns indices of matching IDs.
These arrays do not need to have missing elements, but they must have different orders (otherwise PSI would not be necessary)

## Is your feature request related to a problem?
This will be a first demonstration that PSI can be used to link vertically partitioned data.
The next step is to integrate it into the pipeline and use it to re-order the partitioned datasets

## Additional Context
Part of PSI Epic #2 
Dependent on #13

### PR DESCRIPTION
**Pull request by rsandmann**
## Description
This PR aims to integrate the Python binding of [OpenMined/PSI](https://github.com/OpenMined/PSI) into PyVertical. A new directory `psi` has been added to `src/` where the corresponding files reside. A shell script `build-psi.sh` has been added along with a `WORKSPACE` file to facilitate the build process of PSI using [bazel](https://www.bazel.build/).

Fixes #13 
Fixes #22 

## Affected Dependencies
* [bazel](https://www.bazel.build/)

## How has this been tested?
* Unit tests have been added
* The example code of the [README](https://github.com/rsandmann/PyVertical/tree/psi-integration#usage) has been tested in a Jupyter notebook

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: tests/test_dataloader.py [hunk 1]
```diff
@@ -90,3 +91,17 @@ def test_that_dataset2_must_not_have_data(self):
             dataloader = PartitionDistributingDataLoader(
                 self.dataset1, self.dataset1, batch_size=100
             )
+
```
Comment (TTitcombe) [inline tests/test_dataloader.py:94]: Thanks for the tests. Could we also have tests for `compute_psi` in a `test_psi.py` file

Comment (rsandmann) [reply tests/test_dataloader.py:94]: I added the test but now the GitHub checks fail because the *_psi_bindings.so* file is missing. We probably have to add the `./build-psi.sh` step to the test setup.

**Comment by rsandmann** [comment]
I should add that the example that I added to the README gives an error if `partition_dataset` is called with `remove_data=True`. After discussion with @TTitcombe we decided to go forward with this for now. We should investigate this though, maybe someome with more insight on the matter could have a look at it?

File: tests/test_dataloader.py (hunk 1 above)
Comment (TTitcombe) [reply tests/test_dataloader.py:94]: Because the `PSI` repo uses bazel, i assume we don't need to do any extra setup in the test script? we should be able to call `build-psi.sh`

File: tests/test_psi.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+"""
+Test code in src/psi
+"""
+
+from src.psi.util import compute_psi
+
+def test_compute_psi_returns_correct_indices():
```
Comment (TTitcombe) [inline tests/test_psi.py:7]: Can we use pytest mark_parametrize and include a few edge cases, such as having no intersection

File: tests/test_dataloader.py (hunk 1 above)
Comment (rsandmann) [reply tests/test_dataloader.py:94]: Correct, executing `build-psi.sh` should put the necessary file (*_psi_bindings.so*) in the correct folder, we depend on an existing bazel installation though.

File: tests/test_psi.py (hunk 2 above)
Comment (rsandmann) [reply tests/test_psi.py:7]: That's a good idea! I'll add some more tests

**Comment by TTitcombe** [comment]
@rsandmann looks like the bazel installation in the test script isn't working correctly. If you can't get it to work, remove the psi tests for now. We can get the tests working in a separate PR

File: build-psi.sh [hunk 3]
```diff
@@ -0,0 +1,2 @@
+bazel build @org_openmined_psi//private_set_intersection/python:private_set_intersection &&
```
Comment (TTitcombe) [inline build-psi.sh:1]: We will need to run this script in the github actions test script

**Comment by TTitcombe** [comment]
@rsandmann a small conflict to fix (it looks simple - just keep everything).

The tests pass 🥳 but there's a mypy type checking error caused by `_psi_bindings` import in `src/psi/__init__.py`, I think you can get mypy to ignore this by putting ` # type:ignore` at the end of the `from . import _psi_binding` line
